File: packages/javascript/src/transport.ts
import type { ErrorEvent } from './types'

/**
 * Non-blocking HTTP transport that queues events and flushes periodically.
 * Uses sendBeacon on page unload for reliable delivery.
 */
export class Transport {
  private readonly dsn: string
  private readonly timeout: number
  private readonly maxQueueSize: number
  private readonly flushInterval: number
  private queue: ErrorEvent[] = []
  private timer: ReturnType<typeof setInterval> | null = null

  constructor(
    dsn: string,
    timeout: number = 5000,
    maxQueueSize: number = 100,
    flushIntervalMs: number = 5000,
  ) {
    this.dsn = dsn
    this.timeout = timeout
    this.maxQueueSize = maxQueueSize
    this.flushInterval = flushIntervalMs
    this.startFlushTimer()
    this.installUnloadHandler()
  }

  /** Enqueue an event for async sending. Returns false if queue is full. */
  send(event: ErrorEvent): boolean {
    if (this.queue.length >= this.maxQueueSize) {
      console.warn('nzr_autofix: event queue full, dropping event')
      return false
    }
    this.queue.push(event)

    // Flush immediately if we have 10+ events
    if (this.queue.length >= 10) {
      this.flush()
    }

    return true
  }

  /** Flush all queued events to the server. */
  async flush(): Promise<void> {
    if (this.queue.length === 0) return

    const batch = this.queue.splice(0)

    for (const event of batch) {
      await this.sendWithRetry(event)
    }
  }

  /** Stop the flush timer. */
  destroy(): void {
    if (this.timer) {
      clearInterval(this.timer)
      this.timer = null
    }
  }

  private startFlushTimer(): void {
    if (typeof setInterval === 'undefined') return
    this.timer = setInterval(() => this.flush(), this.flushInterval)
  }

  private installUnloadHandler(): void {
    if (typeof window === 'undefined') return

    const onUnload = (): void => {
      if (this.queue.length === 0) return

      // Use sendBeacon for reliable delivery on page unload
      if (typeof navigator?.sendBeacon === 'function') {
        for (const event of this.queue) {
          const blob = new Blob([JSON.stringify(event)], {
            type: 'application/json',
          })
          navigator.sendBeacon(this.dsn, blob)
        }
        this.queue = []
      }
    }

    window.addEventListener('pagehide', onUnload)
    window.addEventListener('beforeunload', onUnload)
  }

  private async sendWithRetry(
    event: ErrorEvent,
    maxRetries: number = 3,
  ): Promise<void> {
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      try {
        const controller = new AbortController()
        const timeoutId = setTimeout(() => controller.abort(), this.timeout)

        const response = await fetch(this.dsn, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'X-Autofix-DSN': this.dsn,
          },
          body: JSON.stringify(event),
          signal: controller.signal,
        })

        clearTimeout(timeoutId)

        if (response.ok) return

        if (response.status === 429) {
          const retryAfter = parseInt(
            response.headers.get('Retry-After') ?? '5',
            10,
          )
          await sleep(retryAfter * 1000)
          continue
        }

        if (response.status >= 500) {
          await sleep(Math.pow(2, attempt) * 1000)
          continue
        }

        // 4xx (not 429) — don't retry
        return
      } catch {
        await sleep(Math.pow(2, attempt) * 1000)
      }
    }

    console.warn('nzr_autofix: failed to send event after retries')
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

File: packages/python/nzr_autofix/transport.py
"""Background thread transport for sending events (T018)."""

import json
import logging
import queue
import threading
import time
import urllib.request
import urllib.error

logger = logging.getLogger('nzr_autofix')


class Transport:
    """Non-blocking HTTP transport using a background daemon thread."""

    def __init__(self, dsn: str, timeout: float = 5.0, max_queue_size: int = 100):
        self._dsn = dsn
        self._timeout = timeout
        self._queue: queue.Queue = queue.Queue(maxsize=max_queue_size)
        self._worker = threading.Thread(target=self._run, daemon=True)
        self._worker.start()

    def send(self, payload: dict) -> bool:
        """Enqueue a payload for async sending. Returns False if queue is full."""
        try:
            self._queue.put_nowait(payload)
            return True
        except queue.Full:
            logger.warning('nzr_autofix: event queue full, dropping event')
            return False

    def _run(self):
        """Worker loop — drains queue and sends events."""
        while True:
            try:
                payload = self._queue.get(block=True)
                self._send_with_retry(payload)
            except Exception:
                logger.exception('nzr_autofix: transport worker error')

    def _send_with_retry(self, payload: dict, max_retries: int = 3):
        """Send payload with exponential backoff retry."""
        body = json.dumps(payload, default=str).encode('utf-8')

        for attempt in range(max_retries):
            try:
                req = urllib.request.Request(
                    self._dsn,
                    data=body,
                    headers={
                        'Content-Type': 'application/json',
                        'X-Autofix-DSN': self._dsn,
                    },
                    method='POST',
                )
                urllib.request.urlopen(req, timeout=self._timeout)
                return  # success
            except urllib.error.HTTPError as e:
                if e.code == 429:
                    retry_after = int(e.headers.get('Retry-After', 5))
                    time.sleep(retry_after)
                elif e.code >= 500:
                    time.sleep(2 ** attempt)
                else:
                    logger.warning('nzr_autofix: HTTP %d sending event', e.code)
                    return  # don't retry 4xx (except 429)
            except (urllib.error.URLError, OSError):
                time.sleep(2 ** attempt)

        logger.warning('nzr_autofix: failed to send event after %d retries', max_retries)
